#! /bin/sh /usr/share/dpatch/dpatch-run
## 50_g++4.1.dpatch by Rafael Laboissiere <rafael@debian.org>
##
## DP: Allow compilation with g++ 4.1

@DPATCH@

--- octave2.1-2.1.72.orig/liboctave/CMatrix.h
+++ octave2.1-2.1.72/liboctave/CMatrix.h
@@ -204,17 +204,6 @@
 
   ComplexMatrix expm (void) const;
 
-  // column vector by row vector -> matrix operations
-
-  friend ComplexMatrix operator * (const ColumnVector& a,
-				   const ComplexRowVector& b);
-
-  friend ComplexMatrix operator * (const ComplexColumnVector& a,
-				   const RowVector& b);
-
-  friend ComplexMatrix operator * (const ComplexColumnVector& a,
-				   const ComplexRowVector& b);
-
   // matrix by diagonal matrix -> matrix operations
 
   ComplexMatrix& operator += (const DiagMatrix& a);
@@ -290,6 +279,17 @@
 ComplexMatrix Sylvester (const ComplexMatrix&, const ComplexMatrix&,
 			 const ComplexMatrix&);
 
+// column vector by row vector -> matrix operations
+
+ComplexMatrix
+operator * (const ColumnVector& a, const ComplexRowVector& b);
+
+ComplexMatrix
+operator * (const ComplexColumnVector& a, const RowVector& b);
+
+ComplexMatrix
+operator * (const ComplexColumnVector& a, const ComplexRowVector& b);
+
 extern ComplexMatrix operator * (const Matrix&,        const ComplexMatrix&);
 extern ComplexMatrix operator * (const ComplexMatrix&, const Matrix&);
 extern ComplexMatrix operator * (const ComplexMatrix&, const ComplexMatrix&);
--- octave2.1-2.1.72.orig/liboctave/dColVector.h
+++ octave2.1-2.1.72/liboctave/dColVector.h
@@ -103,6 +103,11 @@
   ColumnVector (double *d, int l) : MArray<double> (d, l) { }
 };
 
+// Publish externally used friend functions.
+
+extern ColumnVector real (const ComplexColumnVector& a);
+extern ColumnVector imag (const ComplexColumnVector& a);
+
 MARRAY_FORWARD_DEFS (MArray, ColumnVector, double)
 
 #endif
--- octave2.1-2.1.72.orig/liboctave/dMatrix.h
+++ octave2.1-2.1.72/liboctave/dMatrix.h
@@ -181,10 +181,6 @@
 
   boolMatrix operator ! (void) const;
 
-  // column vector by row vector -> matrix operations
-
-  friend Matrix operator * (const ColumnVector& a, const RowVector& b);
-
   // other operations
 
   Matrix map (d_d_Mapper f) const;
@@ -235,6 +231,15 @@
   Matrix (double *d, int r, int c) : MArray2<double> (d, r, c) { }
 };
 
+// Publish externally used friend functions.
+
+extern Matrix real (const ComplexMatrix& a);
+extern Matrix imag (const ComplexMatrix& a);
+
+// column vector by row vector -> matrix operations
+
+extern Matrix operator * (const ColumnVector& a, const RowVector& b);
+
 extern Matrix Givens (double, double);
 
 extern Matrix Sylvester (const Matrix&, const Matrix&, const Matrix&);
--- octave2.1-2.1.72.orig/liboctave/dNDArray.h
+++ octave2.1-2.1.72/liboctave/dNDArray.h
@@ -131,6 +131,11 @@
   NDArray (double *d, const dim_vector& dv) : MArrayN<double> (d, dv) { }
 };
 
+// Publish externally used friend functions.
+
+extern NDArray real (const ComplexNDArray& a);
+extern NDArray imag (const ComplexNDArray& a);
+
 extern NDArray min (double d, const NDArray& m);
 extern NDArray min (const NDArray& m, double d);
 extern NDArray min (const NDArray& a, const NDArray& b);
--- octave2.1-2.1.72.orig/src/ov.h
+++ octave2.1-2.1.72/src/ov.h
@@ -814,6 +814,15 @@
   octave_value *nil_rep (void) const;
 };
 
+// Publish externally used friend functions.
+
+extern octave_value
+do_unary_op (octave_value::unary_op op, const octave_value& a);
+
+extern octave_value
+do_binary_op (octave_value::binary_op op,
+	      const octave_value& a, const octave_value& b);
+
 #define OV_UNOP_FN(name) \
   inline octave_value \
   name (const octave_value& a) \
